cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -535,6 +540,138 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
     responder.sendJson(HttpResponseStatus.OK, specification);
   }
 
+  @PUT
+  @Path("apps/{app-name}/schedules/{schedule-name}")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void addSchedule(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId,
+                          @PathParam("app-name") String appName,
+                          @PathParam("schedule-name") String scheduleName)
+    throws SchedulerException, BadRequestException, NotFoundException, IOException, AlreadyExistsException {
+    doAddSchedule(request, responder, namespaceId, appName, ApplicationId.DEFAULT_VERSION, scheduleName);
+  }
+
+  @PUT
+  @Path("apps/{app-name}/versions/{app-version}/schedules/{schedule-name}")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void addSchedule(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId,
+                          @PathParam("app-name") String appName,
+                          @PathParam("app-version") String appVersion,
+                          @PathParam("schedule-name") String scheduleName)
+    throws SchedulerException, BadRequestException, NotFoundException, IOException, AlreadyExistsException {
+    doAddSchedule(request, responder, namespaceId, appName, appVersion, scheduleName);
+  }
+
+  private void doAddSchedule(HttpRequest request, HttpResponder responder, String namespaceId, String appId,
+                             String appVersion, String scheduleName)
+    throws IOException, BadRequestException, NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationId applicationId = new ApplicationId(namespaceId, appId, appVersion);
+    ScheduleSpecification scheduleSpecFromRequest;
+    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8)) {
+      // The schedule spec in the request body does not contain the program information
+      scheduleSpecFromRequest = GSON.fromJson(reader, ScheduleSpecification.class);
+    } catch (IOException e) {
+      LOG.debug("Error reading schedule specification from request body", e);
+      throw new IOException("Error reading request body");
```

Comment:
It should at least contain the original exception's message. Otherwise it is not very helpful